File: leetcode1579/solution.cpp
class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        /* Union Find
        
        Time: O(E + V log V) | Space: O(E + V)
        E: Number of edges
        V: Number of vertices
        */

        vector<int> parent_a(n, -1), parent_b(n, -1);
        vector<int> size_a(n, 1), size_b(n, 1);

        set<tuple<int, int, int>> to_remove;

        priority_queue<tuple<int, int, int>> pq_a, pq_b; // this is why there is a V log V in Time Complexity

        for (auto& edge: edges) {
            int type = edge[0];
            int from = edge[1];
            int to = edge[2];

            if (type == 1) {
                pq_a.push({type, from - 1, to - 1});
            } else if (type == 2) {
                pq_b.push({type, from - 1, to - 1});
            } else {
                pq_a.push({type, from - 1, to - 1});
                pq_b.push({type, from - 1, to - 1});
            }
        }

        int remain_a = 0, remain_b = 0;

        while (!pq_a.empty()) {
            auto [type, from, to] = pq_a.top();

            int pf = find(parent_a, from); 
            int pt = find(parent_a, to);

            if (pf == pt) {
                to_remove.insert({type, from, to});
            } else {
                if (size_a[pf] < size_a[pt]) {
                    parent_a[pf] = pt;
                    size_a[pt] += size_a[pf];
                } else {
                    parent_a[pt] = pf;
                    size_a[pf] += size_a[pt];
                }
            }
            pq_a.pop();
        }

        while (!pq_b.empty()) {
            auto [type, from, to] = pq_b.top();

            int pf = find(parent_b, from); 
            int pt = find(parent_b, to);

            if (pf == pt) {
                to_remove.insert({type, from, to});
            } else {
                if (size_b[pf] < size_b[pt]) {
                    parent_b[pf] = pt;
                    size_b[pt] += size_b[pf];
                } else {
                    parent_b[pt] = pf;
                    size_b[pf] += size_b[pt];
                }
            }
            pq_b.pop();
        }

        int self_parent = 0;

        // to check if there is a vertex whose parent is -1
        for (int& p: parent_a) {
            if (p == -1) {
                self_parent += 1;
            }
        }

        for (int& p: parent_b) {
            if (p == -1) {
                self_parent += 1;
            }
        }

        if (self_parent > 2) {
            return -1;
        } else {
            return to_remove.size();
        }
    }

    int find(vector<int>& parent, int current) {
        // find the root
        if (parent[current] == -1) {
            return current;
        } else {
            return find(parent, parent[current]);
        }
    }
};

File: leetcode1579/solution.rs
impl Solution {
    pub fn max_num_edges_to_remove(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        /// Union Find
        /// 
        /// Time: O(E + V) | Space: O(E + V)
        /// where E is the number of edges
        /// and V is the number of nodes
        let n: usize = n as usize;

        let mut size_a: Vec<i32> = vec![1; n];
        let mut size_b: Vec<i32> = vec![1; n];

        let mut parent_a: Vec<i32> = vec![-1; n];
        let mut parent_b: Vec<i32> = vec![-1; n];

        let mut res: i32 = 0;

        for edge in edges.iter() {
            // traverse edges with type 3 first because they are the only ones that Alice and Bob can share
            let tp: i32 = edge[0];
            let x: i32 = edge[1] - 1;
            let y: i32 = edge[2] - 1;

            let px_a: i32 = Self::find(&parent_a, x);
            let py_a: i32 = Self::find(&parent_a, y);
            let px_b: i32 = Self::find(&parent_b, x);
            let py_b: i32 = Self::find(&parent_b, y);

            if tp == 3 {
                if px_a == py_a && px_b == py_b {
                    res += 1;
                } else {
                    if px_a != py_a {
                        Self::unionn(&mut parent_a, &mut size_a, px_a, py_a);
                    }

                    if px_b != py_b {
                        Self::unionn(&mut parent_b, &mut size_b, px_b, py_b);
                    }
                }
            }
        }

        for edge in edges.iter() {
            // traverse edges with type 1 and 2
            let tp: i32 = edge[0];
            let x: i32 = edge[1] - 1;
            let y: i32 = edge[2] - 1;

            let px_a: i32 = Self::find(&parent_a, x);
            let py_a: i32 = Self::find(&parent_a, y);
            let px_b: i32 = Self::find(&parent_b, x);
            let py_b: i32 = Self::find(&parent_b, y);

            if tp == 2 {
                if px_b == py_b {
                    res += 1;
                } else {
                    Self::unionn(&mut parent_b, &mut size_b, px_b, py_b);
                }
            } else if tp == 1 {
                if px_a == py_a {
                    res += 1;
                } else {
                    Self::unionn(&mut parent_a, &mut size_a, px_a, py_a);
                }
            }
        }

        let mut self_parent = 0;

        // to check if there are more than 2 nodes with -1 parent
        // if so, some nodes are not connected
        for p in parent_a {
            if p == -1 {
                self_parent += 1;
            }
        }

        for p in parent_b {
            if p == -1 {
                self_parent += 1;
            }
        }

        return if self_parent > 2 {
            -1
        } else {
            res
        };

        
    }

    fn find(parent: &Vec<i32>, current: i32) -> i32 {
        return if parent[current as usize] == -1 {
                current
            } else {
                Self::find(parent, parent[current as usize])
            }
    }
    
    fn unionn(parent: &mut Vec<i32>, size: &mut Vec<i32>, x: i32, y: i32) {
        let x: usize = x as usize;
        let y: usize = y as usize;
        if size[x] < size[y] {
            parent[x] = y as i32;
            size[y] += size[x];
        } else {
            parent[y] = x as i32;
            size[x] += size[y];
        }
    }
}